src/main/java/com/pinterest/secor/uploader/FormatConvertingUploadManager.java

```diff
@@ -0,0 +1,154 @@
+import com.pinterest.secor.util.ReflectionUtil;
+
+import java.io.File;
+
+import org.apache.hadoop.io.compress.CompressionCodec;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manages uploads using the configured upload manager
+ * Converts the local files to the specified format before upload
+ *
+ * @author Jason Butterfield (jason.butterfield@outreach.io)
+ */
+public class FormatConvertingUploadManager extends UploadManager {
+    private static final Logger LOG = LoggerFactory.getLogger(FormatConvertingUploadManager.class);
+
+    private UploadManager mUploadManager;
+
+    public FormatConvertingUploadManager(SecorConfig config) throws Exception {
+        super(config);
+        mUploadManager = createUploadManager();
+    }
+
+    @Override
+    public TempFileUploadHandle<Handle<?>> upload(LogFilePath localPath) throws Exception {
+        // convert the file from the internal format to the external format
+        System.out.println("localPath: " + localPath);
+        System.out.println("mUploadManager: " + mUploadManager);
+        LogFilePath convertedFilePath = convertFile(localPath);
+
+        // The TempFileUploadHandle will delete the temp file after it resolves
+        return new TempFileUploadHandle(mUploadManager.upload(convertedFilePath), convertedFilePath);
+    }
+
+    /**
+     * This method converts a file 
+     * Exposed for testing only
+     */
+    public LogFilePath convertFile(LogFilePath srcPath) throws Exception {
+        FileReader reader = null;
+        FileWriter writer = null;
+        int copiedMessages = 0;
+
+        String localConvertedPrefix = mConfig.getLocalPath() + '/' +
+            IdUtil.getLocalMessageDir() + "/convertedForUpload";
+        LogFilePath convertedFilePath = srcPath.withPrefix(localConvertedPrefix);
+
+        try {
+            CompressionCodec codec = null;
+            String extension = "";
+            if (mConfig.getCompressionCodec() != null && !mConfig.getCompressionCodec().isEmpty()) {
+                codec = CompressionUtil.createCompressionCodec(mConfig.getCompressionCodec());
+                extension = codec.getDefaultExtension();
+            }
+
+            reader = createReader(srcPath, codec);
+            writer = createWriter(convertedFilePath, codec);
+
+            KeyValue keyVal;
+            while ((keyVal = reader.next()) != null) {
+              writer.write(keyVal);
+              copiedMessages++;
+            }
+        } finally {
+            if (reader != null) {
+                reader.close();
+            }
+            if (writer != null) {
+                writer.close();
+            }
+        }
+
+        LOG.info("converted {} messages from {} to {}",
+            copiedMessages,
+            srcPath.getLogFilePath(),
+            convertedFilePath.getLogFilePath()
+        );
+
+        return convertedFilePath;
+    }
+
+    /**
+     * This method is intended to be overwritten in tests.
```

Comment:
Then you probably want your constructor to take an object that has a method like this defined, so that you can mock that object and replace it using strategy pattern.

Reply:
I copy/pasted the comments here.  overwritten is really a misnomer.  they are intended to be (and are) mocked in the tests.  I'll update the comments.

I agree that passing these into the constructor would be better, but this is the way dependency injection is handled throughout this repo, and adding params to the constructor is not trivial with the way they are loading everything via reflection.